Configured Dockerfile to use poetry, added toml file for poetry config

HughieH opened the pull request:
- My branch is behind main for the Readme file

ntlhui commented on ip-docker/Dockerfile:
```diff
@@ -1,23 +1,43 @@
+    C_INCLUDE_PATH=/usr/include/gdal
+
+# Install Poetry
+RUN pip install poetry
```
invoke pip using `python3.x -m pip` to ensure you are using the correct python

HughieH replied:
updated ths in the Dockerfile, we are using Python 3.9 for this service.

ntlhui commented on ip-docker/Dockerfile:
```diff
@@ -1,23 +1,43 @@
+
+# Install system dependencies required for Poetry and your project
+RUN apt-get update \
+    && apt-get install -y --no-install-recommends \
+       ffmpeg \
+       libsm6 \
+       libxext6 \
+       binutils \
+       libproj-dev \
+       gdal-bin \
+       libgdal-dev \
+       curl
```
Install `python3-pip` to ensure pip exists.

ntlhui commented on ip-docker/Dockerfile:
```diff
@@ -1,23 +1,43 @@
+ENV PATH="${PATH}:/root/.poetry/bin"
+
+# Copy only pyproject.toml and poetry.lock (if available) to cache dependencies
+COPY pyproject.toml poetry.lock* /ip_service/
```
`poetry.lock` should always exist!

HughieH replied:
added poetry.lock to the repo

ntlhui commented on ip-docker/Dockerfile:
```diff
@@ -1,23 +1,43 @@
+# Install Poetry
+RUN pip install poetry
+
+# Ensure that Poetry is in the PATH
+ENV PATH="${PATH}:/root/.poetry/bin"
```
This smells suspicious - why do you need it?

HughieH replied:
honestly did not need this, removed it from Dockerfile

ntlhui commented on ip-docker/pyproject.toml:
```diff
@@ -0,0 +1,31 @@
+
+[tool.poetry.dependencies]
+python = "^3.9"
+Flask = "2.3.1"
+requests = "2.28.2"
+boto3 = "1.26.120"
+numpy = "1.24.3"
+rasterio = "1.3.6"
+rioxarray = "0.14.1"
+xarray = "2023.4.2"
+distancerasters = "0.3.2"
+opencv-contrib-python = "4.7.0.72"
+dask = "2023.4.0"
+simplejson = "3.19.1"
+Flask-Cors = "3.0.10"
+flask-sock = "0.6.0"
+h11 = "0.14.0"
+websockets = "11.0.2"
+geopandas = "0.12.2"
+pandas = "1.5.3"
```
This smells like a `pip freeze` dump!  Are these the top level requirements of your package?

HughieH replied:
As mentioned in my comment for the classification service. Our main goal is just to get all our services connected right now. I'm temporarily using the same dependencies as stated in requirements.txt. Dependencies and packages will be updated once we push towards integrating everything and making sure the websocket routes work.

ntlhui commented on ip-docker/pyproject.toml:
```diff
@@ -0,0 +1,31 @@
+[tool.poetry]
+name = "MM-ip-service"
+version = "0.1.0"
+description = "IP service for MM team in E4E"
+authors = []
```
You seem to be missing package information!

ntlhui commented on ip-docker/pyproject.toml:
You seem to be missing entry point information!
You seem to be missing a lock file!

HughieH replied:
Added both of these.

HughieH commented:
## Updates about IP service:
- Resolved all issues with the Dockerfile and poetry toml file, also added the poetry.lock file.
- Changed file structure of our repo to follow Python package conventions. We also now use a poetry entrypoint for out Docker container. 
- Updated imports for the "mm_image_processing/util" directory to use relative imports. Was running into "module not found" errors. I believe this is to do with running poetry as a module, and thus when we use poetry as our entrypoint it doesn't run server.py as a script.

ntlhui commented on mm_image_processing/server.py:
```diff
@@ -9,16 +9,18 @@
 
+# Ultility functions, using relative imports here since we are using a poetry entrypoint to run our Flask server
+# Using absolute imports results in Module not found errors whenever we try to call functions from the util directory
+from .util import image_processor
+from .util.s3_img_getter import getImg
+from .util.image_processor import processImgFromLocal
+from .util.s3_img_getter import getImg, deleteImg
```
The way this repo is structured, your package provides the following modules: `mm_image_processing` and `util`.

I would recommend that the `util` module be a submodule of `mm_image_processing`.  In order for any of these to be proper python modules, `util` must also have a `__init__.py`, and be referenced as `mm_image_processing.util`

HughieH replied:
Updated to follow the above format.

ntlhui commented on requirements.txt:
Why does this still exist if you are using `poetry`?

HughieH replied:
deleting this

ntlhui commented on mm_image_processing/.gitignore:
This should live in the root of the repository.  Nominally, this will have been generated from https://gitignore.io.

HughieH replied:
Removed this gitignore.

ntlhui commented on pyproject.toml:
```diff
@@ -0,0 +1,36 @@
+[tool.poetry.dependencies]
+python = "^3.9"
+Flask = "2.3.1"
+requests = "2.28.2"
```
`requests` is not directly invoked from your code!

HughieH replied:
removed this

ntlhui commented on pyproject.toml:
```diff
@@ -0,0 +1,36 @@
+Flask = "2.3.1"
+requests = "2.28.2"
+boto3 = "1.26.120"
+numpy = "1.24.3"
```
This is extremely limiting, are you sure? Did you mean `^1.24`?

ntlhui commented on pyproject.toml:
```diff
@@ -0,0 +1,36 @@
+
+[tool.poetry.dependencies]
+python = "^3.9"
+Flask = "2.3.1"
```
This is extremely limiting, are you sure? Did you mean `^2.3`?

HughieH replied:
Updated

ntlhui commented on pyproject.toml:
```diff
@@ -0,0 +1,36 @@
+rioxarray = "0.14.1"
+xarray = "2023.4.2"
+distancerasters = "0.3.2"
+opencv-contrib-python = "4.7.0.72"
```
Are we actually using extra/contrib modules from OpenCV?

ntlhui commented on pyproject.toml:
```diff
@@ -0,0 +1,36 @@
+xarray = "2023.4.2"
+distancerasters = "0.3.2"
+opencv-contrib-python = "4.7.0.72"
+dask = "2023.4.0"
```
`dask` is not imported!

HughieH replied:
removed dask dependancy

ntlhui commented on pyproject.toml:
```diff
@@ -0,0 +1,36 @@
+distancerasters = "0.3.2"
+opencv-contrib-python = "4.7.0.72"
+dask = "2023.4.0"
+simplejson = "3.19.1"
```
`simplejson` is not imported!

HughieH replied:
removed this

ntlhui commented on pyproject.toml:
```diff
@@ -0,0 +1,36 @@
+opencv-contrib-python = "4.7.0.72"
+dask = "2023.4.0"
+simplejson = "3.19.1"
+Flask-Cors = "3.0.10"
```
This seems very limiting! Did you mean `^3.0`?

HughieH replied:
updated this

ntlhui commented on pyproject.toml:
```diff
@@ -0,0 +1,36 @@
+dask = "2023.4.0"
+simplejson = "3.19.1"
+Flask-Cors = "3.0.10"
+flask-sock = "0.6.0"
```
This seems very limiting!  do you mean `^0.6`?

ntlhui commented on pyproject.toml:
```diff
@@ -0,0 +1,36 @@
+simplejson = "3.19.1"
+Flask-Cors = "3.0.10"
+flask-sock = "0.6.0"
+h11 = "0.14.0"
```
`h11` is not imported!

This smells suspicious - why do you need http1.1?

HughieH replied:
removed this

ntlhui commented on pyproject.toml:
```diff
@@ -0,0 +1,36 @@
+Flask-Cors = "3.0.10"
+flask-sock = "0.6.0"
+h11 = "0.14.0"
+websockets = "11.0.2"
```
This is very limiting - did you mean `^11.0`?

ntlhui commented on pyproject.toml:
```diff
@@ -0,0 +1,36 @@
+flask-sock = "0.6.0"
+h11 = "0.14.0"
+websockets = "11.0.2"
+geopandas = "0.12.2"
```
This is very limiting - did you mean `^0.12`?

Also, latest `geopandas` is `0.14.3`.

ntlhui commented on pyproject.toml:
```diff
@@ -0,0 +1,36 @@
+h11 = "0.14.0"
+websockets = "11.0.2"
+geopandas = "0.12.2"
+pandas = "1.5.3"
```
`pandas` is a dependency of `geopandas` - are you explicitly using `pandas` in this project?

`1.5.3` is very limiting - did you mean `^1.5`?

HughieH replied:
I don't believe so, i'll remove it for now and if we do end up requiring pandas in the future we can re-add it to the toml file.

ntlhui commented on pyproject.toml:
In general, this smells of using `pip freeze` to generate the dependency list.  This is not acceptable practice.

HughieH replied:
Will be updating dependencies and making sure we actally use the appropriate packages.

ntlhui commented on pyproject.toml:
```diff
@@ -0,0 +1,36 @@
+[tool.poetry]
+name = "mm_image_processing"
+version = "0.1.0"
+packages = [] # add required pacakges later?
```
This should probably explicitly list `mm_image_processing` as a module provided by this package

HughieH replied:
Included this.

ntlhui commented on mm_image_processing/util/image_processor.py:
```diff
@@ -8,7 +8,7 @@
 import rasterio
 import rioxarray
-from util import features
+from . import features # relative import for util.features
```
See `utils` import

HughieH replied:
updated.

ntlhui commented on Dockerfile:
```diff
@@ -0,0 +1,34 @@
+EXPOSE 8080
+
+# poetry entrypoint to run the service
+ENTRYPOINT ["poetry", "run", "image-processing-server"]
```
Since you've installed using poetry into the system python, you should be able to invoke `image-processing-server` as an executable.

HughieH replied:
Updated to using the exe for our entrypoint.

ntlhui commented on Dockerfile:
```diff
@@ -0,0 +1,34 @@
+# FOR ARM64, 
+FROM --platform=linux/amd64 python:3.9-slim-buster
```
Are you aware that this is Debian 10 (current is Debian 12)?

HughieH replied:
I've updated this to using 3.9-slim-bookworm, it uses debian 12 and is actively maintained.

ntlhui commented on Dockerfile:
```diff
@@ -0,0 +1,34 @@
+# Install system dependencies
+RUN apt-get update \
+    && apt-get install -y --no-install-recommends \
+       ffmpeg \
+       libsm6 \
+       libxext6 \
+       binutils \
+       libproj-dev \
+       gdal-bin \
+       libgdal-dev \
+       curl
```
This smells funny.  What is the justification for each of these?  Are they actually being used?

HughieH replied:
slimmed this down, includid python3-pip and only stuff needed for rasterio

ntlhui commented on pyproject.toml:
```diff
@@ -0,0 +1,36 @@
+[tool.poetry]
+name = "mm_image_processing"
+version = "0.1.0"
```
There are now two sources of truth! here and `mm_image_processing/__init__.py`.  This is not ideal

HughieH replied:
Updated names.

ntlhui commented on pyproject.toml:
```diff
@@ -0,0 +1,36 @@
+version = "0.1.0"
+packages = [] # add required pacakges later?
+description = "IP service for MM team in E4E"
+authors = ["E4E - Mangrove Monitoring SWE Team"]
```
List authors individually, see `git log` for all names

HughieH commented:
## Updates IP-Service:
* Added carel operator for poetry dependancies, also removed un-needed libraries
* updated Dockerfile sys dependancies

ntlhui commented on pyproject.toml:
```diff
@@ -0,0 +1,37 @@
+    'Christopher L. Crutchfield <ccrutchf@ucsd.edu>',
+    'Zach Lawrence <zclawrence@ucsd.edu>',
+    'Hou Wai Wan <h1wan@ucsd.edu>',
+    'Leon Lin',
+    'Devin Muzzy'
```
Ideally all authors will have emails, but this is not critical.

ntlhui commented on mm_image_processing/util/image_processor.py:
```diff
@@ -5,15 +5,16 @@
+import logging
 import rasterio
 import rioxarray
-from util import features
+
+from rasterio.plot import reshape_as_image, reshape_as_raster
 from rasterio.features import shapes
-import geopandas as gpd
-import numpy as np
+from mm_image_processing.util import features 
 
-import logging
```
Note that PEP8 suggests the following ordering:

```
import logging

import geopandas as gpd
import numpy as np
import rasterio
import rioxarray
from rasterio.features import shapes
from rasterio.plot import reshape_as_image, reshape_as_raster

from mm_image_processing.util import features
```

HughieH replied:
updated to PEP8 as well

ntlhui commented on mm_image_processing/server.py:
```diff
@@ -9,16 +9,17 @@
 import asyncio
 import websockets
-import util.image_processor as image_processor
 
 from threading import Thread, Lock
-from util.s3_img_getter import getImg
 from flask_sock import Sock
 from flask import Flask, request, make_response
 from flask_cors import CORS
-from util.image_processor import processImgFromLocal
-from util.s3_img_getter import getImg, deleteImg
 
+# Ultility functions from the util directory
+from mm_image_processing.util import image_processor
+from mm_image_processing.util.s3_img_getter import getImg
+from mm_image_processing.util.image_processor import processImgFromLocal
+from mm_image_processing.util.s3_img_getter import getImg, deleteImg
```
Note that PEP8 recommends the following:

```
import asyncio
import json
import logging
import pickle  # only needed for testing without classification service
import random
import string
from threading import Lock, Thread

import requests
import websockets
from flask import Flask, make_response, request
from flask_cors import CORS
from flask_sock import Sock

# Ultility functions from the util directory
from mm_image_processing.util import image_processor
from mm_image_processing.util.image_processor import processImgFromLocal
from mm_image_processing.util.s3_img_getter import deleteImg, getImg
```

Also, the following are not used:
- `random`
- `string`
- `threading.Lock`
- `threading.Thread`
- `requests`
- `flask.make_response`
- `flask.request`
- `mm_image_processing.util.s3_img_getter.deleteImg`

HughieH replied:
updated to PEP8 standards

ntlhui commented on README.md:
```diff
@@ -1,15 +1,42 @@
\ No newline at end of file
+"docker run -p {HOST_MACHINE_PORT}:5000 --net=mmict_bridge ip_service"
+
+# Libraries and Dependencies:
+
+Below is the preiouvs requirements.txt file:
+(Still sorting through what will be required for our IP service)
+
+Flask==2.3.1
+requests==2.28.2
+boto3==1.26.120
+numpy==1.24.3
+rasterio==1.3.6
+rioxarray==0.14.1
+xarray==2023.4.2
+distancerasters==0.3.2
+opencv-contrib-python==4.7.0.72
+dask==2023.4.0
+simplejson==3.19.1
+Flask-Cors==3.0.10
+flask-sock==0.6.0 
+h11==0.14.0 
+websockets==11.0.2
+geopandas==0.12.2
+pandas==1.5.3
+
```
All of this should be in git history - you can probably get rid of this section

HughieH replied:
deleted past req.txt

ntlhui reviewed (approved):
See additional nonessential changes
